dir.c

```diff
@@ -622,8 +622,17 @@ static void add_pattern_to_hashsets(struct pattern_list *pl, struct path_pattern
 	if (!pl->use_cone_patterns)
 		return;
 
-	if (!strcmp(given->pattern, "/*"))
+	if (given->flags & PATTERN_FLAG_NEGATIVE &&
+	    given->flags & PATTERN_FLAG_MUSTBEDIR &&
+	    !strcmp(given->pattern, "/*")) {
+		pl->full_cone = 0;
```

Comment:
Apologies if we already discussed this (I'm having trouble finding it if we did), but is there somewhere that git enforces that the sparse-checkout file is "valid", or are the entries guaranteed to be in a particular order?

I ask because I'm wondering if after setting `pl->full_cone = 0` here we might read an entry that sets it to 1 below (or vice-versa).  Is that the expected behavior?

Reply:
The order is absolutely important in the `sparse-checkout` file. In our typical cone-based patterns, we start with these two lines:

```
/*     # include all files
!/*/   # except those in folders
```

The order gives the meaning to the negation. The last pattern to match the file `/A/file.txt` is `!/*/`, so the sparse-checkout says "do not include this file". (If no pattern matches at all, the default is "do not include this file".)

If they are reversed:

```
!/*/     # do not include things in folders
/*       # include everything
```

then the sparse-checkout file will match everything, as the _last_ pattern to match is a positive pattern.

So, this code is saying "If I see `!/*/`, then set `full_cone` to zero. If I see `/*`, then set `full_cone` to true."

If there are other patterns that make this file not look like a cone-based file, then `use_cone_patterns` is set to zero and `full_cone` is ignored.